### ISSUE TITLE
xgboost doesn't build under cuda 10.1

### ISSUE THREAD

**Issue by rongou** [dmlc/xgboost#4220]
Looks like some deprecated functions have been removed in CUDA 10.1.

@RAMitchell @canonizer 

```bash
/home/rou/src/xgboost/src/tree/updater_gpu.cu(143): error: identifier "__shfl_up" is undefined
          detected during instantiation of "void xgboost::tree::ReduceScanByKey(xgboost::common::Span<xgboost::GradientPair, -1L>, xgboost::common::Span<xgboost::GradientPair, -1L>, xgboost::common::Span<xgboost::GradientPair, -1L>, xgboost::common::Span<const xgboost::tree::NodeIdT, -1L>, xgboost::common::Span<const xgboost::tree::NodeIdT, -1L>, int, int, int, xgboost::common::Span<xgboost::GradientPair, -1L>, xgboost::common::Span<int, -1L>, xgboost::common::Span<const xgboost::tree::NodeIdT, -1L>, xgboost::tree::NodeIdT) [with BLKDIM_L1L3=256, BLKDIM_L2=512]" 
(679): here

1 error detected in the compilation of "/tmp/tmpxft_000047ff_00000000-6_updater_gpu.cpp1.ii".
CMake Error at gpuxgboost_generated_updater_gpu.cu.o.Release.cmake:279 (message):
  Error generating file
  /home/rou/src/xgboost/build/CMakeFiles/gpuxgboost.dir/src/tree/./gpuxgboost_generated_updater_gpu.cu.o


CMakeFiles/gpuxgboost.dir/build.make:9460: recipe for target 'CMakeFiles/gpuxgboost.dir/src/tree/gpuxgboost_generated_updater_gpu.cu.o' failed
make[2]: *** [CMakeFiles/gpuxgboost.dir/src/tree/gpuxgboost_generated_updater_gpu.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/rou/src/xgboost/src/common/host_device_vector.cu: In instantiation of ‘void xgboost::HostDeviceVectorImpl<T>::DeviceShard::Init(xgboost::HostDeviceVectorImpl<T>*, int) [with T = long unsigned int]’:
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:1:   required from ‘xgboost::HostDeviceVectorImpl<T>::InitShards()::<lambda(int, xgboost::HostDeviceVectorImpl<T>::DeviceShard&)> [with T = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:13:   required from ‘struct xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = long unsigned int]::<lambda(int, struct xgboost::HostDeviceVectorImpl<long unsigned int>::DeviceShard&)>’
/home/rou/src/xgboost/src/common/host_device_vector.cu:204:23:   required from ‘void xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:171:11:   required from ‘xgboost::HostDeviceVectorImpl<T>::HostDeviceVectorImpl(size_t, T, xgboost::GPUDistribution) [with T = long unsigned int; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:430:12:   required from ‘xgboost::HostDeviceVector<T>::HostDeviceVector(size_t, T, xgboost::GPUDistribution) [with T = long unsigned int; size_t = long unsigned int]’
/home/rou/src/xgboost/include/xgboost/data.h:186:14:   required from here
/home/rou/src/xgboost/src/common/host_device_vector.cu:56:16: error: cannot call member function ‘size_t xgboost::HostDeviceVectorImpl<T>::Size() const [with T = long unsigned int; size_t = long unsigned int]’ without object
       LazyResize(vec_->Size());
            ~~~~^~
/home/rou/src/xgboost/src/common/host_device_vector.cu: In instantiation of ‘void xgboost::HostDeviceVectorImpl<T>::DeviceShard::Init(xgboost::HostDeviceVectorImpl<T>*, int) [with T = xgboost::Entry]’:
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:1:   required from ‘xgboost::HostDeviceVectorImpl<T>::InitShards()::<lambda(int, xgboost::HostDeviceVectorImpl<T>::DeviceShard&)> [with T = xgboost::Entry]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:13:   required from ‘struct xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = xgboost::Entry]::<lambda(int, struct xgboost::HostDeviceVectorImpl<xgboost::Entry>::DeviceShard&)>’
/home/rou/src/xgboost/src/common/host_device_vector.cu:204:23:   required from ‘void xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = xgboost::Entry]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:171:11:   required from ‘xgboost::HostDeviceVectorImpl<T>::HostDeviceVectorImpl(size_t, T, xgboost::GPUDistribution) [with T = xgboost::Entry; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:430:12:   required from ‘xgboost::HostDeviceVector<T>::HostDeviceVector(size_t, T, xgboost::GPUDistribution) [with T = xgboost::Entry; size_t = long unsigned int]’
/home/rou/src/xgboost/include/xgboost/data.h:186:14:   required from here
/home/rou/src/xgboost/src/common/host_device_vector.cu:56:16: error: cannot call member function ‘size_t xgboost::HostDeviceVectorImpl<T>::Size() const [with T = xgboost::Entry; size_t = long unsigned int]’ without object
/home/rou/src/xgboost/src/common/host_device_vector.cu: In instantiation of ‘void xgboost::HostDeviceVectorImpl<T>::DeviceShard::Init(xgboost::HostDeviceVectorImpl<T>*, int) [with T = float]’:
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:1:   required from ‘xgboost::HostDeviceVectorImpl<T>::InitShards()::<lambda(int, xgboost::HostDeviceVectorImpl<T>::DeviceShard&)> [with T = float]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:13:   required from ‘struct xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = float]::<lambda(int, struct xgboost::HostDeviceVectorImpl<float>::DeviceShard&)>’
/home/rou/src/xgboost/src/common/host_device_vector.cu:204:23:   required from ‘void xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = float]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:171:11:   required from ‘xgboost::HostDeviceVectorImpl<T>::HostDeviceVectorImpl(size_t, T, xgboost::GPUDistribution) [with T = float; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:430:12:   required from ‘xgboost::HostDeviceVector<T>::HostDeviceVector(size_t, T, xgboost::GPUDistribution) [with T = float; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:593:16:   required from here
/home/rou/src/xgboost/src/common/host_device_vector.cu:56:16: error: cannot call member function ‘size_t xgboost::HostDeviceVectorImpl<T>::Size() const [with T = float; size_t = long unsigned int]’ without object
/home/rou/src/xgboost/src/common/host_device_vector.cu: In instantiation of ‘void xgboost::HostDeviceVectorImpl<T>::DeviceShard::Init(xgboost::HostDeviceVectorImpl<T>*, int) [with T = xgboost::detail::GradientPairInternal<float>]’:
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:1:   required from ‘xgboost::HostDeviceVectorImpl<T>::InitShards()::<lambda(int, xgboost::HostDeviceVectorImpl<T>::DeviceShard&)> [with T = xgboost::detail::GradientPairInternal<float>]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:13:   required from ‘struct xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = xgboost::detail::GradientPairInternal<float>]::<lambda(int, struct xgboost::HostDeviceVectorImpl<xgboost::detail::GradientPairInternal<float> >::DeviceShard&)>’
/home/rou/src/xgboost/src/common/host_device_vector.cu:204:23:   required from ‘void xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = xgboost::detail::GradientPairInternal<float>]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:171:11:   required from ‘xgboost::HostDeviceVectorImpl<T>::HostDeviceVectorImpl(size_t, T, xgboost::GPUDistribution) [with T = xgboost::detail::GradientPairInternal<float>; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:430:12:   required from ‘xgboost::HostDeviceVector<T>::HostDeviceVector(size_t, T, xgboost::GPUDistribution) [with T = xgboost::detail::GradientPairInternal<float>; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:594:16:   required from here
/home/rou/src/xgboost/src/common/host_device_vector.cu:56:16: error: cannot call member function ‘size_t xgboost::HostDeviceVectorImpl<T>::Size() const [with T = xgboost::detail::GradientPairInternal<float>; size_t = long unsigned int]’ without object
/home/rou/src/xgboost/src/common/host_device_vector.cu: In instantiation of ‘void xgboost::HostDeviceVectorImpl<T>::DeviceShard::Init(xgboost::HostDeviceVectorImpl<T>*, int) [with T = int]’:
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:1:   required from ‘xgboost::HostDeviceVectorImpl<T>::InitShards()::<lambda(int, xgboost::HostDeviceVectorImpl<T>::DeviceShard&)> [with T = int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:205:13:   required from ‘struct xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = int]::<lambda(int, struct xgboost::HostDeviceVectorImpl<int>::DeviceShard&)>’
/home/rou/src/xgboost/src/common/host_device_vector.cu:204:23:   required from ‘void xgboost::HostDeviceVectorImpl<T>::InitShards() [with T = int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:171:11:   required from ‘xgboost::HostDeviceVectorImpl<T>::HostDeviceVectorImpl(size_t, T, xgboost::GPUDistribution) [with T = int; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:430:12:   required from ‘xgboost::HostDeviceVector<T>::HostDeviceVector(size_t, T, xgboost::GPUDistribution) [with T = int; size_t = long unsigned int]’
/home/rou/src/xgboost/src/common/host_device_vector.cu:595:16:   required from here
/home/rou/src/xgboost/src/common/host_device_vector.cu:56:16: error: cannot call member function ‘size_t xgboost::HostDeviceVectorImpl<T>::Size() const [with T = int; size_t = long unsigned int]’ without object
CMake Error at gpuxgboost_generated_host_device_vector.cu.o.Release.cmake:279 (message):
  Error generating file
  /home/rou/src/xgboost/build/CMakeFiles/gpuxgboost.dir/src/common/./gpuxgboost_generated_host_device_vector.cu.o


CMakeFiles/gpuxgboost.dir/build.make:2434: recipe for target 'CMakeFiles/gpuxgboost.dir/src/common/gpuxgboost_generated_host_device_vector.cu.o' failed
make[2]: *** [CMakeFiles/gpuxgboost.dir/src/common/gpuxgboost_generated_host_device_vector.cu.o] Error 1
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/gpuxgboost.dir/all' failed
make[1]: *** [CMakeFiles/gpuxgboost.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by hcho3** [comment]
@rongou We are building the PyPI package (https://pypi.org/project/xgboost/#files) using CUDA 8.0. Can a machine with CUDA 10.1 run code compiled with older version of CUDA?

**Comment by rongou** [comment]
Probably yes for older gpu cards, but for newer cards (volta, turing) they may need newer cuda versions.

**Comment by hcho3** [comment]
@rongou So there is no backward compatibility, i.e. you cannot use older binaries to run on newer cards?

**Comment by RAMitchell** [comment]
Older binaries should run on newer cards, although they may need to JIT compile ptx code resulting in some start-up overhead. @rongou do you want to send a PR enclosing the relevant cuda API calls inside a __CUDA_ARCH__ macro?

**Comment by pisymbol** [comment]
Same issue for me. I was rebuilding my container for CUDA 10.1 and boom, xgboost fails to build. Same trace as above. Ubuntu 18.04 if that matters.

**Comment by hcho3** [comment]
@RAMitchell I'm preparing a re-vamping of the CI pipeline, now that we have better funding for it. In particular, we should de-couple build tests (does it compile?) from unit tests (does it run correctly on a given target)?

**Comment by pisymbol** [comment]
Is there a time table on when CUDA 10.1 will be supported by xgboost?

**Comment by hcho3** [comment]
@pisymbol Until the problem is addressed, would you be able to use the pre-compiled binary from PyPI?

**Comment by hcho3** [comment]
@pisymbol Allow us some time, since CUDA 10.1 has been release only a week ago. In the future, we'll set up a system to respond to new CUDA releases in a more timely manner.

**Comment by pisymbol** [comment]
Yeah, I understand your pain. Really, I do. I am doing exactly that and bypassing my CUDA enabled build for the PyPi one. That should hold me over for now.

I know CUDA is such a pain. :-)

**Comment by hcho3** [comment]
Fixed by #4223.
